registration/include/pcl/registration/impl/correspondence_estimation.hpp

```diff
@@ -126,45 +127,86 @@ CorrespondenceEstimation<PointSource, PointTarget, Scalar>::determineCorresponde
 
   std::vector<int> index(1);
   std::vector<float> distance(1);
-  pcl::Correspondence corr;
-  unsigned int nr_valid_correspondences = 0;
+  std::vector<pcl::Correspondences> per_thread_correspondences(num_threads_);
+  for(auto& corrs: per_thread_correspondences) {
+    corrs.reserve(2 * indices_->size() / num_threads_);
+  }
 
   // Check if the template types are the same. If true, avoid a copy.
   // Both point types MUST be registered using the POINT_CLOUD_REGISTER_POINT_STRUCT
   // macro!
   if (isSamePointType<PointSource, PointTarget>()) {
     // Iterate over the input set of source indices
-    for (const auto& idx : (*indices_)) {
+    #pragma omp parallel for \
+      default(none) \
+      shared(tree_, indices_, max_dist_sqr, per_thread_correspondences) \
+      firstprivate(index, distance) \
+      num_threads(num_threads_)
+    for (int i = 0; i < static_cast<int>(indices_->size()); i++) {
+      const auto& idx = (*indices_)[i];
       tree_->nearestKSearch((*input_)[idx], 1, index, distance);
       if (distance[0] > max_dist_sqr)
         continue;
 
+      pcl::Correspondence corr;
       corr.index_query = idx;
       corr.index_match = index[0];
       corr.distance = distance[0];
-      correspondences[nr_valid_correspondences++] = corr;
+      per_thread_correspondences[omp_get_thread_num()].push_back(corr);
```

Comment:
```suggestion
      per_thread_correspondences[omp_get_thread_num()].emplace_back(std::move(corr));
```

And same below. The compiler might not remove the forced copy in L155 (It might move `corr` but `push_back` has 1 compulsory copy in it)

Reply:
Done